plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,303 @@
+    const ip = event.properties?.['$ip']
+    const host = event.properties?.['$host']
+    const timezone = event.properties?.['$timezone'] || TIMEZONE_FALLBACK
+    const timestampMs = DateTime.fromISO(timestamp).toMillis()
+    const teamId = event.team_id
+
+    return { userAgent, ip, host, timezone, timestampMs, teamId }
+}
+
+const localSaltMap: Record<string, Uint32Array> = {}
+const mutex = new ConcurrencyController(1)
+
+export async function getSaltForDay(db: DB, timestamp: number, timeZone: string): Promise<Uint32Array> {
+    // get the day based on the timezone
+    const parts = new Intl.DateTimeFormat('en', {
+        timeZone,
+        year: 'numeric',
+        month: '2-digit',
+        day: '2-digit',
+    }).formatToParts(new Date(timestamp))
+    const year = parts.find((part) => part.type === 'year')?.value
+    const month = parts.find((part) => part.type === 'month')?.value
+    const day = parts.find((part) => part.type === 'day')?.value
+    const yyyymmdd = `${year}-${month}-${day}`
+
+    // see if we have it locally
+    if (localSaltMap[yyyymmdd]) {
+        return localSaltMap[yyyymmdd]
+    }
+
+    // get the salt for the day from redis, but only do this once for this node process
+    return mutex.run({
+        fn: async (): Promise<Uint32Array> => {
+            // check if we got the salt while waiting for the mutex
+            if (localSaltMap[yyyymmdd]) {
+                return localSaltMap[yyyymmdd]
+            }
+
+            // try to get it from redis instead
+            const saltBase64 = await db.redisGet<string | null>(
+                `cookieless_salt:${yyyymmdd}`,
+                null,
+                'cookielessServerHashStep'
+            )
+            if (saltBase64) {
+                const salt = base64StringToUint32Array(saltBase64)
+                localSaltMap[yyyymmdd] = salt
+                return salt
+            }
+
+            // looks like we'll need to create a new one, but check the date is within range
+            if (!isCalendarDateValid(yyyymmdd)) {
+                throw new Error('Date is out of range')
+            }
+
+            // try to write a new one to redis, but don't overwrite
+            const newSaltParts = createRandomUint32x4()
+            const setResult = await db.redisSetNX(
+                `cookieless_salt:${yyyymmdd}`,
+                uint32ArrayToBase64String(newSaltParts),
+                'cookielessServerHashStep',
+                SALT_TTL_SECONDS
+            )
+            if (setResult === 'OK') {
+                localSaltMap[yyyymmdd] = newSaltParts
+                return newSaltParts
+            }
+
+            // if we couldn't write, it means that it exists in redis already
+            const saltBase64Retry = await db.redisGet<string | null>(
+                `cookieless_salt:${yyyymmdd}`,
+                null,
+                'cookielessServerHashStep'
+            )
```

Comment:
I came here ready to complain about the possible race conditions, but couldn't find a way to trigger them, so lgtm :)

Reply:
I did think about this a fair bit, especially as this looks a bit like the well-known double-checked locking bug in Java, but with that said I am definitely glad that you couldn't find one :)